deepchem/models/tensorgraph/tensor_graph.py

```diff
@@ -0,0 +1,339 @@
+    features = self.features[0]
+    if not self.built:
+      self.build()
+    with self.tensor_objects['Graph'].as_default():
+      saver = tf.train.Saver()
+      with tf.Session() as sess:
+        saver.restore(sess, self.last_checkpoint)
+        fetches = [x.out_tensor for x in self.outputs]
+        feed_dict = {features.out_tensor: X}
+        fetched_values = sess.run(fetches, feed_dict=feed_dict)
+        return np.array(fetched_values)
+
+  def predict_proba_on_batch(self, X):
+    if not self.built:
+      self.build()
+    with self.tensor_objects['Graph'].as_default():
+      saver = tf.train.Saver()
+      with tf.Session() as sess:
+        saver.restore(sess, self.last_checkpoint)
+        out_tensors = [x.out_tensor for x in self.outputs]
+        fetches = out_tensors
+        feed_dict = self._construct_feed_dict(X, None, None, None)
+        fetched_values = sess.run(fetches, feed_dict=feed_dict)
+        return np.array(fetched_values)
+
+  def topsort(self):
+    return nx.topological_sort(self.nxgraph)
+
+  def build(self):
+    with self._get_tf("Graph").as_default():
+      order = self.topsort()
+      for node in order:
+        node_layer = self.layers[node]
+        parents = self.parents[node]
+        with tf.name_scope(node):
+          node_layer.__call__(*parents)
+      self.built = True
+
+    for layer in self.layers.values():
+      if layer.tensorboard:
+        self.tensorboard = True
+    tf.summary.scalar("loss", self.loss.out_tensor)
+    for layer in self.layers.values():
+      if layer.tensorboard:
+        tf.summary.tensor_summary(layer.name, layer.out_tensor)
+    if self.tensorboard:
+      writer = self._get_tf("FileWriter")
+      writer.add_graph(self._get_tf("Graph"))
+      writer.close()
+
+  def set_loss(self, layer):
+    self.loss = layer
+
+  def add_label(self, layer):
+    self.labels.append(layer)
+
+  def add_feature(self, layer):
+    self.features.append(layer)
+
+  def add_output(self, layer):
+    self.outputs.append(layer)
+
+  def save(self):
+    # Remove out_tensor from the object to be pickled
+    must_restore = False
+    out_tensors = []
+    tensor_objects = self.tensor_objects
+    self.tensor_objects = {}
+    if self.built:
+      must_restore = True
+      out_tensors = []
+      for node in self.topsort():
+        node_layer = self.layers[node]
+        out_tensors.append(node_layer.out_tensor)
+        node_layer.out_tensor = None
+      self.built = False
+
+    # Pickle itself
+    pickle_name = os.path.join(self.model_dir, "model.pickle")
+    with open(pickle_name, 'wb') as fout:
+      pickle.dump(self, fout)
+
+    # add out_tensor back to everyone
+    if must_restore:
+      for index, node in enumerate(self.topsort()):
+        node_layer = self.layers[node]
+        node_layer.out_tensor = out_tensors[index]
+      self.built = True
+    self.tensor_objects = tensor_objects
+
+  def _get_tf(self, obj):
```

Comment:
I'm a little uncertain about having an internal parameter dictionary. I can see why it's convenient, but there's also a risk of having it balloon with unwanted objects in subclasses. This is more a soft feeling than a hard request though. Perhaps more docs could help control the dictionary.